### PR TITLE
Certificate pinning (EXPOSUREAPP-10635)

### PR DESCRIPTION
**Pull request by BMItr**
This PR addresses Exposureapp-10635. Subtask for the TicketValidation Story.
Implements server certificate check.
Test added.

### CODE REVIEW

**Review by jurajkusnier — COMMENTED** [review]
How do you get list of server certificates? I don't see any OkHttpClient/Retrofit implementation when this should be implemented?

**Comment by BMItr** [comment]
> How do you get list of server certificates? I don't see any OkHttpClient/Retrofit implementation when this should be implemented?

Check the validate fun from here: https://github.com/corona-warn-app/cwa-app-android/blob/e62f6771fe9fae959d0334faa2[…]ki/coronawarnapp/dccticketing/core/server/DccTicketingServer.kt
I use the handshake.peercertificates of the raw response.

File: Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    <!-- XBUT: DCC Ticketing: Consent One: Cancel Dialog: Continue Button  -->
+    <string name="dcc_ticketing_consent_one_cancel_dialog_continue_btn">"Überprüfung Fortsetzen"</string>
+    <!-- XBUT: DCC Ticketing: Consent One: Cancel Dialog: Cancel Button  -->
+    <string name="dcc_ticketing_consent_one_cancel_dialog_cancel_btn">"Überprüfung Abbrechen"</string>
+
+    <!-- Not in Dcc ticketing allow list dialog-->
+    <!-- XTXT: not in AllowList dialog title -->
+    <string name="dcc_ticketing_not_in_allow_list_dialog_title">Sicherheitshinweis</string>
+    <!-- XTXT: not in AllowList dialog message -->
+    <string name="dcc_ticketing_not_in_allow_list_dialog_message">Der Anbieter ist uns nicht bekannt. Bitte überprüfen Sie, ob der QR-Code wirklich von dem Anbieter stammt und ob Sie ihm Ihr Zertifikat zur Verfügung stellen wollen.</string>
```
Comment (KathrinLuecke) [inline Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml:57]: This text was changed. It is now: "Der Anbieter ist uns nicht bekannt. Aus Sicherheitsgründen ist die Überprüfung bei Ticketbuchungen nur für bekannte Anbieter möglich. Bitte kontaktieren Sie den Anbieter, der Ihnen diesen QR-Code bereitgestellt hat."
See https://www.figma.com/file/KuDEcMSREg3VhZRhC3FXTY/COVID19_Master_2.11?node-id=37230%3A26423
Please adapt.

**Review by KathrinLuecke — CHANGES_REQUESTED** [review]
Please change dcc_ticketing_not_in_allow_list_dialog_message

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/check/DccTicketingServerCertificateChecker.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        if (requiredJwkSet.isEmpty()) {
+            Timber.d("Didn't find jwk for required kid, aborting")
+            throw DccTicketingServerCertificateCheckException(ErrorCode.CERT_PIN_NO_JWK_FOR_KID)
+        }
+
+        return requiredJwkSet
+    }
+
+    private fun DccJWK.toCertificate(): Certificate = x5c.first()
```
Comment (jurajkusnier) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/check/DccTicketingServerCertificateChecker.kt:81]: could you use or move the implementation into https://github.com/corona-warn-app/cwa-app-android/blob/release/2.15.x/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/core/common/DccJWKExt.kt as something very similar is already there

**Comment by jurajkusnier** [comment]
> 

Ok, got it, I was looking for `socketFactory` or `X509TrustManager` implementation. So we have to call this manually in every okhttp/retrofit response

**Comment by BMItr** [comment]
due rebase issues, PR will be closed. (follow up will come)

File: Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml (hunk 1 above)
Comment (BMItr) [reply Corona-Warn-App/src/main/res/values-de/dcc_ticketing_strings.xml]: unrelated.

**Pull request closed without merging**